diff -ru ../vsftpd-3.0.2/logging.c vsftpd-3.0.2/logging.c
--- ../vsftpd-3.0.2/logging.c	2013-06-13 19:20:41.000000000 +0800
+++ vsftpd-3.0.2/logging.c	2023-01-11 18:35:40.263199496 +0800
@@ -14,6 +14,12 @@
 #include "sysstr.h"
 #include "session.h"
 
+#define VSFTP_FIFO_PATH  "/run/vsftpd.fifo"
+#define MSG_LEN 256
+#include <fcntl.h>
+#include <string.h>
+
+
 /* File local functions */
 static int vsf_log_type_is_transfer(enum EVSFLogEntryType type);
 static void vsf_log_common(struct vsf_session* p_sess, int succeeded,
@@ -30,6 +36,7 @@
 void
 vsf_log_init(struct vsf_session* p_sess)
 {
+  p_sess->vsftpd_fifo_fd = open(VSFTP_FIFO_PATH, O_WRONLY);
   if (tunable_syslog_enable || tunable_tcp_wrappers)
   {
     vsf_sysutil_openlog(1);
@@ -71,6 +78,17 @@
   }
 }
 
+extern void vsf_log_upload_finish(struct vsf_session* p_sess)
+{
+    int retval = vsf_sysutil_lock_file_write(p_sess->vsftpd_fifo_fd);
+    if (vsf_sysutil_retval_is_error(retval))
+        return;
+    char buff[MSG_LEN] = {0};
+    memcpy(buff, str_getbuf(&p_sess->log_str), MSG_LEN);
+    vsf_sysutil_write_loop(p_sess->vsftpd_fifo_fd, buff, MSG_LEN);
+    vsf_sysutil_unlock_file(p_sess->vsftpd_fifo_fd);
+}
+
 static int
 vsf_log_type_is_transfer(enum EVSFLogEntryType type)
 {
diff -ru ../vsftpd-3.0.2/main.c vsftpd-3.0.2/main.c
--- ../vsftpd-3.0.2/main.c	2013-06-13 19:20:41.000000000 +0800
+++ vsftpd-3.0.2/main.c	2022-11-01 21:33:02.386748988 +0800
@@ -54,7 +54,7 @@
     /* Pre-chroot() cache */
     INIT_MYSTR, INIT_MYSTR, INIT_MYSTR, INIT_MYSTR, 1,
     /* Logging */
-    -1, -1, INIT_MYSTR, 0, 0, 0, INIT_MYSTR, 0,
+    -1, -1, -1, INIT_MYSTR, 0, 0, 0, INIT_MYSTR, 0,
     /* Buffers */
     INIT_MYSTR, INIT_MYSTR,
     /* Parent <-> child comms */
diff -ru ../vsftpd-3.0.2/postlogin.c vsftpd-3.0.2/postlogin.c
--- ../vsftpd-3.0.2/postlogin.c	2013-06-13 19:20:41.000000000 +0800
+++ vsftpd-3.0.2/postlogin.c	2023-01-11 18:09:25.725959023 +0800
@@ -27,6 +27,7 @@
 #include "ssl.h"
 #include "vsftpver.h"
 #include "opts.h"
+#include <string.h>
 
 /* Private local functions */
 static void handle_pwd(struct vsf_session* p_sess);
@@ -80,6 +81,9 @@
 static int data_transfer_checks_ok(struct vsf_session* p_sess);
 static void resolve_tilde(struct mystr* p_str, struct vsf_session* p_sess);
 
+// LD: add
+extern void vsf_log_upload_finish(struct vsf_session* p_sess);
+
 void
 process_post_login(struct vsf_session* p_sess)
 {
@@ -1134,6 +1138,8 @@
   if (trans_ret.retval == 0)
   {
     success = 1;
+    // LD: add
+    vsf_log_upload_finish(p_sess);
     vsf_log_do_log(p_sess, 1);
   }
   if (trans_ret.retval == -1)
diff -ru ../vsftpd-3.0.2/session.h vsftpd-3.0.2/session.h
--- ../vsftpd-3.0.2/session.h	2013-06-13 19:20:41.000000000 +0800
+++ vsftpd-3.0.2/session.h	2022-11-01 21:33:02.386748988 +0800
@@ -68,6 +68,7 @@
   /* Logging related details */
   int xferlog_fd;
   int vsftpd_log_fd;
+  int vsftpd_fifo_fd;
   struct mystr remote_ip_str;
   unsigned long log_type;
   long log_start_sec;
